openstack/workflow/v2/workflows/requests.go

```diff
@@ -0,0 +1,56 @@
+package workflows
+
+import (
+	"io"
+
+	"github.com/gophercloud/gophercloud"
+)
+
+// CreateOptsBuilder allows extension to add additional parameters to the Create request.
+type CreateOptsBuilder interface {
+	ToWorkflowCreateQuery() (io.Reader, string, error)
+}
+
+// CreateOpts specifies parameters used to create a cron trigger.
+type CreateOpts struct {
+	// Scope is the scope of the workflow.
+	// Allowed values are "private" and "public".
+	Scope string `q:"scope"`
+
+	// Namespace will define the namespace of the workflow.
+	Namespace string `q:"namespace"`
+
+	// Definition is the workflow definition written in Mistral Workflow Language v2.
+	Definition io.Reader
+}
+
+// ToWorkflowCreateQuery constructs a request query string from CreateOpts.
+func (opts CreateOpts) ToWorkflowCreateQuery() (io.Reader, string, error) {
+	q, err := gophercloud.BuildQueryString(opts)
+	return opts.Definition, q.String(), err
+}
+
+// Create requests the creation of a new execution.
+func Create(client *gophercloud.ServiceClient, opts CreateOptsBuilder) (r CreateResult) {
+	url := createURL(client)
+	var b io.Reader
+	if opts != nil {
+		tmpB, query, err := opts.ToWorkflowCreateQuery()
+		if err != nil {
+			r.Err = err
+			return
+		}
+		url += query
+		b = tmpB
+	}
```

Comment:
Can this instead be:

```go
if opts != nil {
  query, err := opts.ToWorkflowCreatQuery()
  if err != nil {
    r.Err = err
    return
  }
  url += query
  b = opts.Definition
}
```

Or is there a reason to have `opts.Definition` come from `ToWorkflowCreateQuery()`?

Reply:
I based this from what I saw in objectstorage package, because workflows have the same way of working regarding the create process: the POST request takes a definition file as body (text/plain), and other parameters are sent through URL query.

This is how it's handled on objectstorage:
https://github.com/gophercloud/gophercloud/blob/master/openstack/objectstorage/v1/objects/requests.go#L171 

I just kept the same signature. I'll probably rename the function "ToWorkflowCreateQuery" to "ToWorkflowCreateParams" to be less ambiguous.

We can't make your proposal as we have a CreateOptsBuilder which is an interface, so no access to opts.Definition.

Reply:
> I based this from what I saw in objectstorage package, because workflows have the same way of working regarding the create process: the POST request takes a definition file as body (text/plain), and other parameters are sent through URL query.

Right, and re-using existing patterns is definitely the way to go. At least as a start.

What I was pointing out is that you aren't actually doing anything with `Definition` inside `ToWorkflowCreateParams`, though. For example, in the `objects` package, the `Content` is used to create the etag hash.

My question was more _will_ you ever be doing anything with `Definition`? 

However,

> We can't make your proposal as we have a CreateOptsBuilder which is an interface, so no access to opts.Definition.

Oh, right - duh. I totally overlooked that. Busy day :)

Reply:
> What do you think about it?

Nah, that'll be a large break from the current patterns. Let's leave it as-is.